### DISCUSSION TITLE
Procedures vs Macros

### DISCUSSION THREAD

**Discussion by andrewgoz** [TiddlyWiki/TiddlyWiki5#7673] [Q&A]
It appears recent TiddlyWiki changes are suggesting the use of procedures over macros. The tiddler linked below needs to parse text dates like "6-Sep-2022" into a Unix timestamp like "1662393600000". My \define macro works just how I'd like, but when I tried it as a procedure it doesn't work.

The second points of each list group are showing the working \define. The first two numbers should be the same and the difference between them is one is obtained via my preferred option of direct field transclusion {{!!released}}, while the second number is variable substitution, which I'd prefer to not do.

http://www.scss.com.au/family/andrew/tiddlywiki/#ApexChart%20Timeline%20Test%20JSON

The documentation indicates that double angle brackets are how parameters are used, but something clearly isn't working.

I spent over an hour last night on another issue trying various brackets, transclusions, macros before finally getting the magic incantation that worked. This has to be the most frustrating aspect of working with TiddlyWiki :-(

**Comment by andrewgoz** [comment]
I've solved my parsing problem, but the issue of parameter handling is still a mystery. I'll stick with \define for now.

**Comment by pmario** [comment]
In [ApexChart - Area Chart - Basic](http://www.scss.com.au/family/andrew/tiddlywiki/#ApexChart%20-%20Area%20Chart%20-%20Basic) you use a lot of `<$text` widgets to convert the filtered transclusion back to text

```
<$link/>

<$apexchart type=area title="Fundamental Analysis of Stocks" height=350>
{
          "series": [{
          "name": "STOCK ABC",
          "data": <$text text={{{ [[ApexChart Series Data]get[text]jsonextract[monthDataSeries1],[prices]] }}}/>
```

The easier way would be to use the `\function` pragma.

```
\function Data() [[ApexChart Series Data]get[text]jsonextract[monthDataSeries1],[prices]]

<$link/>

<$apexchart type=area title="Fundamental Analysis of Stocks" height=350>
{
          "series": [{
          "name": "STOCK ABC",
          "data": <<Data>>
```

**Reply by pmario** [reply]
Also see:  https://talk.tiddlywiki.org/t/would-it-be-possible-to-write-a-filter-operator-that-outputs-the-result-of-a-filter-as-plain-text/7775

**Comment by pmario** [comment]
I would add the CodeMirror plugin with a basic configuration, which makes editing the code much easier, since CM allows tabs to be used for indentation. 

My basic code example would look as follows. IMO it's much easier to read for new users. And more advanced users who need more parameters would need to read the Apex docs. If they understand this doc they should be able to adapt the JSON structure accordingly. 

```
\function Labels() [[ApexChart Series Data]get[text]jsonextract[monthDataSeries1],[dates]]
\function Data() [[ApexChart Series Data]get[text]jsonextract[monthDataSeries1],[prices]]

<$link/>

<$apexchart type=area title="Fundamental Analysis of Stocks" height=350>
{
	"series": [{
		"name": "STOCK ABC",
		"data": <<Data>>
	}],
	"chart": {
		"zoom": {"enabled": false}
	},
	"dataLabels": {"enabled": false},
	"stroke": {"curve": "straight"},
	"title": {"align": "left"},
	"subtitle": {
		"text": "Price Movements",
		"align": "left"
	},
	"labels": <<Labels>>,
	"xaxis": {"type": "datetime"},
	"yaxis": {"opposite": true},
	"legend": {"horizontalAlign": "left"}
}
</$apexchart>
```

Here's a ZIP with a configuration I would use.

[test-apex.zip](https://github.com/Jermolene/TiddlyWiki5/files/12331642/test-apex.zip)

**Reply by pmario** [reply]
![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/c9737d42-ae5c-4150-8415-655bb2de65c6)

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/ef421f89-55a5-4fc2-8ed1-fcac0f3dad46)

**Reply by pmario** [reply]
The CodeMirror setting will allow you to adjust the tab-width that is used in the code editor. I did set it to 4 in the wiki

**Comment by andrewgoz** [comment]
Thanks for your suggestions. I'm still new to this \function business. In some ways it looks better, but it does have the disadvantage of separating those instructions away from all the other chart options. A problem if the chart widget is in the middle of a pile of WikiText, but not a problem if it's a tiddler by itself and transcluded. Which is the "TiddlyWiki way" anyway.

I'm sure CodeMirror is a nice plugin, but the charts plugin is big enough without recommending another relatively large plugin. I think a manual tidy of the JSON will help - I've just been smashing the copy-and-paste from the ApexCharts website then doing the minimal to make it JSON-friendly.

I don't know how most people will want to use my plugin anyway. I suspect most won't go too far beyond minimal options changes and a data dictionary tiddler! I really don't know.

The primary objective of re-implementing the ApexCharts demos is to stress-test my plugin. I still have a lot to do.

**Reply by pmario** [reply]
I personally would not directly call the `<$apexchart` widget in the content text anyway. I would abstract it away with a `<<myChart title>>` procedure and only expose a minimum number of parameters to the user. like so: 

The following code should work. But it seems your widget has a problem to parse parameters in the right way. So it shows `[object Object]`
This is an initialisation problem of the widget. You should **not** parse parameters yourself. You need to use the core functions. ... But I'm not sure there. Did not have a closer look at the code. 

```
\procedure myChart(title)
\function Labels() [[ApexChart Series Data]get[text]jsonextract[monthDataSeries1],[dates]]
\function Data() [[ApexChart Series Data]get[text]jsonextract[monthDataSeries1],[prices]]

<$apexchart type=area title=<<title>> height=350>
{
	"series": [{
		"name": "STOCK ABC",
		"data": <<Data>>
	}],
	"chart": {
		"zoom": {"enabled": false}
	},
	"dataLabels": {"enabled": false},
	"stroke": {"curve": "straight"},
	"title": {"align": "left"},
	"subtitle": {
		"text": "Price Movements",
		"align": "left"
	},
	"labels": <<Labels>>,
	"xaxis": {"type": "datetime"},
	"yaxis": {"opposite": true},
	"legend": {"horizontalAlign": "left"}
}
</$apexchart>
\end

<$link/>

<<myChart title:"Fundamental Analysis of Stocks">>
```

**Reply by andrewgoz** [reply]
Thanks for pointing out the attribute parsing issue. I think I've got it right now...

Maybe it's a style difference, but I was thinking rather than wrapping the widget in a macro, but instead put the widget, related functions, etc into their own tiddler, then transclude where needed?

**Reply by pmario** [reply]
>Maybe it's a style difference, but I was thinking rather than wrapping the widget in a macro, but instead put the widget, related functions, etc into their own tiddler, then transclude where needed?

I'm not 100% sure. I think I'd start with a macro definition as pointed out. 

Then may be change to a "full" transclusion, where every widget + config + meta-data  is in 1 tiddler which is transcluded into the content tiddler. So everything is together and no or low noise in the content tiddler.

**Reply by pmario** [reply]
We do have the new "parametrized transclusions" now. I do not have much experience with it. I did not use it yet. It may be an interesting experiment to be used with your widget. -- future will tell
